### PR TITLE
support netpoll in dev branch

### PR DESCRIPTION
**Pull request by cch123**
mosn/easygo@27b5582

add field to desc, avoid calling file.Fd which will set the conn to blocking mode

remove fd copy in NewDesc, prevent pre close file during GC, before the connection actually close

call ReadOnce again until readAgain == false, if there is still data to read when TLS is enabled.

clear read timeout timer before doRead, set timer after doRead to enable/disable trigger for read timeout event to send heartbeat request

### CODE REVIEW

**Comment by cch123** [comment]
todo
test heartbeat...

File: pkg/network/connection.go [hunk 1]
```diff
@@ -109,6 +109,7 @@ type connection struct {
 	eventLoop        *eventLoop
 	ev               *connEvent
 	readTimeoutTimer *utils.Timer
+	readBufferMux    sync.Mutex
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:112]: 四个都封装到一个结构体吧

File: pkg/network/connection.go [hunk 2]
```diff
@@ -204,16 +205,32 @@ func (c *connection) attachEventLoop(lctx context.Context) {
 
 	// create a new timer and bind it to connection
 	c.readTimeoutTimer = utils.NewTimer(buffer.ConnReadTimeout, func() {
+		c.readBufferMux.Lock()
+		defer c.readBufferMux.Unlock()
+
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:210]: 确定Onevent加锁没问题？

Comment (cch123) [reply pkg/network/connection.go:210]: 这个就比较尴尬了，OnEvent 里一定不能用 readbuffer。。。

Comment (cch123) [reply pkg/network/connection.go]: 目前两个 ReadTimeout 里都是往连接里写数据，暂时没用 readBuffer，还好，我先把临界区改小了

File: pkg/network/connection.go [hunk 3]
```diff
@@ -239,14 +299,16 @@ func (c *connection) attachEventLoop(lctx context.Context) {
 		},
 
 		onHup: func() bool {
-			log.DefaultLogger.Errorf("[network] [event loop] [onHup] ReadHup error. Connection = %d, Remote Address = %s", c.id, c.RemoteAddr().String())
+			log.DefaultLogger.Errorf("[network] [event loop] [onHup] ReadHup error. Connection = %d, Remote Address = %s", c.id, c.RemoteAddr().String(), c.file.Fd())
```
Comment (wangfakang) [inline pkg/network/connection.go:302]: loss format for `c.file.Fd()` .

Comment (cch123) [reply pkg/network/connection.go]: fixed, this onHup func will be removed in future,
currently it's never called

File: pkg/mtls/crypto/tls/conn.go [hunk 4]
```diff
@@ -26,6 +26,10 @@ var outBufPool = sync.Pool{
 	},
 }
 
+func (c *Conn) HasMoreData() bool {
```
Comment (taoyuanyuan) [inline pkg/mtls/crypto/tls/conn.go:29]: add comment to mtls/crypto/README

Comment (taoyuanyuan) [inline pkg/mtls/crypto/tls/conn.go:29]: 统一放到transfer_tls.go 文件吧，然后可以给这个文件换个名字

File: pkg/network/eventloop.go [hunk 5]
```diff
@@ -32,191 +33,107 @@ var (
...
-	readPool  = mosnsync.NewWorkerPool(runtime.NumCPU())
-	writePool = mosnsync.NewWorkerPool(runtime.NumCPU())
+	readPool  = mosnsync.NewWorkerPool(4 * runtime.NumCPU())
 
-	rrCounter                 uint32
-	poolSize                  uint32 = 1 //uint32(runtime.NumCPU())
-	eventLoopPool                    = make([]*eventLoop, poolSize)
-	errEventAlreadyRegistered        = errors.New("event already registered")
+	rrCounter     uint32
+	poolSize      uint32 = 1 //uint32(runtime.NumCPU())
```
Comment (ballontt) [inline pkg/network/eventloop.go:39]: 请问这里为什么是设置了1，而不是uint32(runtime.NumCPU())

Comment (cch123) [reply pkg/network/eventloop.go:39]: 之前在 4c (GOMAXPROCS=4)环境下做了压测，开多个 reactor 还不如只有一个，具体数据看看这个
https://github.com/mosn/mosn/pull/1511#issuecomment-747310113

**Comment by taoyuanyuan** [comment]
modify mtls/crypto/README

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Review by nejisama — APPROVED** [review]

**Pull request merged**